### ISSUE TITLE
Weird false positive from `useless_attribute` lint

### ISSUE THREAD

**Issue by sgrif** [rust-lang/rust-clippy#2910]
This occurred with Clippy 0.0.195 on rustc 1.27.0-nightly (ac3c2288f 2018-04-18). The following code triggered the `useless_attribute` lint on the `derive` line even though it clearly should not be:

```rust
#[derive(Deserialize, Default)]
#[serde(deny_unknown_fields)]
pub struct Config {
    #[serde(default)]
    pub print_schema: PrintSchema,
}
```

Seen in this travis build for Diesel: https://travis-ci.org/diesel-rs/diesel/jobs/399697117#L616

**Comment by mati865** [comment]
Seems to work with latest Clippy: [playgound link](https://play.rust-lang.org/?gist=94ebbc9c5c02446bfc64a4a911afe14c&version=stable&mode=debug&edition=2015)

**Comment by sgrif** [comment]
I'm actually not convinced it's related to the clippy version -- I can't replicate the failure locally. We also did not change the clippy version used or the nightly used, the travis build just started failing from this for no apparent reason. The span for the lint also looks completely wrong, as it's not a full attribute.

**Comment by oli-obk** [comment]
Did you bump the serde version? Maybe it's proc macro related

**Comment by sgrif** [comment]
Ah, you're right. `cargo update` led to `    Updating serde_derive v1.0.66 -> v1.0.70` and I'm now able to reproduce locally. All that said, I'm not sure where to open this bug. I don't think that repro script shows that this doesn't occur on the latest clippy version, we have plenty of other `#[derive(Deserialize)]` that work just fine, only this one spot is causing the lint to fail. The `proc_macro2` version did not change.

**Comment by oli-obk** [comment]
Wonderful! I'll investigate what changed in those serde versions

**Comment by mikerite** [comment]
This might be unrelated but it looks like the build uses two different versions of `proc-macro2`.
```
   Compiling proc-macro2 v0.4.6
   Compiling proc-macro2 v0.3.8
```

**Comment by sgrif** [comment]
That was true before this issue occurred

**Comment by sgrif** [comment]
One thing that didn't get explicitly mentioned that I just noticed, Clippy thinks that `derive` is a lint here.

**Comment by vessd** [comment]
I got the same issue.
rustc 1.29.0-nightly (64f7de921 2018-07-12)
clippy https://github.com/rust-lang-nursery/rust-clippy.git#06d67101

cargo log: https://pastebin.com/Qjmv06iJ

Only `derive` with `serde` are affected.

**Comment by JayKickliter** [comment]
Is there any advice on dealing with this? I just started seeing this on a codebase that hasn't changed in a few weeks.

**Comment by mariosangiorgio** [comment]
I'm seeing the same issue. In my case it's with the `FromPrimitive` and `ToPrimitive` from `num-derive` (version 0.2.3)

**Comment by akavel** [comment]
Also seeing the issue with `FromPrimitive`, in [akavel/nixme &rarr; src/lib.rs l.61](https://github.com/akavel/nixme/blob/8e5946da6483cdae7a48ce95b2fc0dae61bbb304/src/lib.rs#L61) (warning: newbie code!)

```
error: useless lint attribute
  --> src\lib.rs:61:1
   |
61 | #[derive(FromPrimitive)]
   | ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: #[deny(clippy::useless_attribute)] on by default
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/master/index.html#useless_attribute
```

**Comment by phansch** [comment]
I just managed to reproduce this using the `num_derive` crate and hope to have a fix PR up tomorrow.
